## Importación de paquetes

In [151]:
import os
import numpy as np
import pandas as pd
np.set_printoptions(
    suppress=True,
    linewidth=100,
    precision=2
)

## Obtención de la ruta del script

In [152]:
os.getcwd()

'c:\\Users\\ernesto.molina\\OneDrive - FUNKEN\\Documentos\\FKTrabajo\\Internos\\Curso Data Analyst\\S28 - A Loan Data Example with NumPy'

## Importación de los datos

In [153]:
data_name = 'loan-data.csv' # Nombre del archivo que contiene los datos a procesar

In [154]:
raw_data_np = np.genfromtxt(data_name, delimiter=';', dtype='str')
raw_data_np[:10,:]

array([['id', 'issue_d', 'loan_amnt', 'loan_status', 'funded_amnt', 'term', 'int_rate',
        'installment', 'grade', 'sub_grade', 'verification_status', 'url', 'addr_state',
        'total_pymnt'],
       ['48010226', 'May-15', '35000.0', 'Current', '35000.0', ' 36 months', '13.33', '1184.86',
        'C', 'C3', 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226', 'CA', '9452.96'],
       ['57693261', '', '30000.0', 'Current', '30000.0', ' 36 months', 'þëè.89', '938.57', 'A',
        'A5', 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', 'NY', '4679.7'],
       ['59432726', 'Sep-15', '15000.0', 'Current', '15000.0', ' 36 months', 'íîå.53', '494.86',
        'B', 'B5', 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', 'PA', '1969.83'],
       ['53222800', 'Jul-15', '9600.0', 'Current', '9600.0', ' 36 months', 'þëè.89', '300.35',
        'A', 'A5', 'Not

Los datos originales se encuentran organizados de la siguiente manera:
- La primera fila del arreglo contiene las etiquetas que identifican cada una de las variables registradas en la base de datos.
- Las filas subsecuentes contienen los datos asociados a cada instancia o registro. Cada registro debe contar con un identificador (ID) único que garantice su correcta diferenciación dentro del conjunto de datos.

In [155]:
raw_data_np = raw_data_np.transpose()
raw_data_np

array([['id', '48010226', '57693261', ..., '50415990', '46154151', '66055249'],
       ['issue_d', 'May-15', '', ..., 'Jun-15', 'Apr-15', 'Dec-15'],
       ['loan_amnt', '35000.0', '30000.0', ..., '10000.0', '', '10000.0'],
       ...,
       ['url', 'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', ...,
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249'],
       ['addr_state', 'CA', 'NY', ..., 'CA', 'OH', 'IL'],
       ['total_pymnt', '9452.96', '4679.7', ..., '2185.64', '3199.4', '301.9']],
      shape=(14, 10001), dtype='<U69')

Los datos se reorganizan de manera que cada elemento del arreglo corresponda a una de las columnas del conjunto de datos original. Esta transformación se realiza con el propósito de facilitar la identificación y el análisis de posibles valores erróneos, inconsistentes o atípicos presentes en los registros.

## Segmentación de los Datos Crudos por Categoría

In [156]:
raw_data_copy = raw_data_np.copy()
caracteristicas_data = {}
for i in range(len(raw_data_copy[:,1:])):
    valores_unicos = np.unique(raw_data_copy[i,1:], return_counts=True)
    contenedor_temporal = []
    dict_temporal = {}
    for j in range(0,(len(valores_unicos)),2):
        contenedor_temporal = (valores_unicos[j], valores_unicos[j+1])
        for k in range(len(contenedor_temporal[0])):
            dict_temporal[contenedor_temporal[0][k]] = contenedor_temporal[1][k]
    caracteristicas_data[raw_data_copy[i,0]] = dict_temporal
label_data = [key for key, _ in caracteristicas_data.items()]
label_data

[np.str_('id'),
 np.str_('issue_d'),
 np.str_('loan_amnt'),
 np.str_('loan_status'),
 np.str_('funded_amnt'),
 np.str_('term'),
 np.str_('int_rate'),
 np.str_('installment'),
 np.str_('grade'),
 np.str_('sub_grade'),
 np.str_('verification_status'),
 np.str_('url'),
 np.str_('addr_state'),
 np.str_('total_pymnt')]

A partir de esta etapa se obtienen dos estructuras de datos:
- label_data: Contiene las etiquetas asociadas a cada una de las columnas de la base de datos.
- caracteristicas_data: Diccionario que resume las características de cada columna, incluyendo los valores únicos que la componen y la frecuencia de ocurrencia de cada uno de ellos. Esta información resulta útil para el análisis exploratorio de datos y la detección de posibles inconsistencias en los registros.

## Exploración de los valores por categoría

### Categoría 'id'

Esta categoría corresponde a los identificadores únicos asignados a cada registro de préstamo. 

Los valores contenidos en esta columna actúan como claves de identificación, garantizando la unicidad de cada observación dentro de la base de datos.

In [164]:
#caracteristicas_data[label_data[0]]
list(caracteristicas_data[label_data[0]].items())[:10]

[(np.str_('12606806'), np.int64(1)),
 (np.str_('13026045'), np.int64(1)),
 (np.str_('1312426'), np.int64(1)),
 (np.str_('14187903'), np.int64(1)),
 (np.str_('14428496'), np.int64(1)),
 (np.str_('14470171'), np.int64(1)),
 (np.str_('14550531'), np.int64(1)),
 (np.str_('14669896'), np.int64(1)),
 (np.str_('15389420'), np.int64(1)),
 (np.str_('16130931'), np.int64(1))]

In [8]:
arreglo_count = [item for _, item in caracteristicas_data[label_data[0]].items()]
max(arreglo_count)

np.int64(1)

El análisis de los identificadores presentes en la categoría 'id' evidencia que:

- Mantienen una correspondencia única con cada uno de los registros extraídos, por lo que pueden considerarse claves únicas dentro del conjunto de datos.
- Se descarta la presencia de registros duplicados, ya que cada identificador presenta una única ocurrencia en la base de datos analizada.
- Los identificadores proporcionados siguen una codificación consistente basada en un serial numérico. No obstante, debido a que su función es exclusivamente identificativa y no representan una variable cuantitativa, se considera apropiado preservarlos en formato de cadena de caracteres durante las etapas posteriores del procesamiento de datos.

### Categoría 'issue_d'

Esta categoría corresponde al mes en el que el préstamo fue financiado. Su contenido permite identificar el período temporal asociado a la formalización y desembolso de cada préstamo.

In [171]:
#caracteristicas_data[label_data[1]]
list(caracteristicas_data[label_data[1]].items())[:10]

[(np.str_(''), np.int64(500)),
 (np.str_('Apr-15'), np.int64(757)),
 (np.str_('Aug-15'), np.int64(846)),
 (np.str_('Dec-15'), np.int64(997)),
 (np.str_('Feb-15'), np.int64(530)),
 (np.str_('Jan-15'), np.int64(770)),
 (np.str_('Jul-15'), np.int64(1061)),
 (np.str_('Jun-15'), np.int64(654)),
 (np.str_('Mar-15'), np.int64(559)),
 (np.str_('May-15'), np.int64(741))]

Del análisis de los datos contenidos en la categoría 'issue_d' se observa lo siguiente:

- Los registros proporcionados corresponden a préstamos otorgados a lo largo del año 2015.
- No se identifican etiquetas diferentes que hagan referencia a un mismo mes.
- Se detectan 500 registros para los cuales el mes de otorgamiento no fue almacenado correctamente, apareciendo en la base de datos como una cadena vacía ('').

Recomendación: reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior dentro de las estructuras de datos de NumPy.

### Categoría 'loan_amnt'

 Esta categoría representa el monto del préstamo solicitado por el prestatario. En caso de que el departamento de crédito haya realizado una reducción del monto inicialmente solicitado durante el proceso de evaluación, el valor registrado en esta variable corresponde al monto ajustado y finalmente aprobado.

In [172]:
#caracteristicas_data[label_data[2]]
list(caracteristicas_data[label_data[2]].items())[:10]

[(np.str_(''), np.int64(500)),
 (np.str_('1000.0'), np.int64(30)),
 (np.str_('10000.0'), np.int64(628)),
 (np.str_('10025.0'), np.int64(1)),
 (np.str_('10050.0'), np.int64(3)),
 (np.str_('10075.0'), np.int64(1)),
 (np.str_('10100.0'), np.int64(6)),
 (np.str_('10150.0'), np.int64(2)),
 (np.str_('10175.0'), np.int64(1)),
 (np.str_('10200.0'), np.int64(13))]

Del análisis de los datos contenidos en la categoría 'loan_amnt' se observa lo siguiente:

- Los registros proporcionados corresponden al monto de los préstamos asociados a cada solicitud contenida en la base de datos.
- Los valores almacenados representan cantidades numéricas y, por tanto, pueden ser tratados como datos de tipo flotante para facilitar su análisis y procesamiento.
- Se detectan 500 registros en los que el monto del préstamo no fue almacenado correctamente, encontrándose representado mediante una cadena vacía ('').
- No se identificaron valores con formatos atípicos o inconsistentes en comparación con el resto de los registros de esta categoría, lo que indica una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.
- Convertir los valores válidos de la categoría desde cadenas de texto a un tipo de dato numérico (por ejemplo, float), permitiendo la aplicación de técnicas estadísticas y de análisis cuantitativo sobre esta variable.

### Categoría 'loan_status'

Categoría que contiene el estado actual del préstamo realizado.

In [ ]:
#caracteristicas_data[label_data[3]]
list(caracteristicas_data[label_data[3]].items())[:10]

{np.str_(''): np.int64(500),
 np.str_('Charged Off'): np.int64(61),
 np.str_('Current'): np.int64(8538),
 np.str_('Default'): np.int64(11),
 np.str_('Fully Paid'): np.int64(505),
 np.str_('In Grace Period'): np.int64(56),
 np.str_('Issued'): np.int64(184),
 np.str_('Late (16-30 days)'): np.int64(30),
 np.str_('Late (31-120 days)'): np.int64(115)}

Las categorías identificadas son las siguientes:

- Fully Paid: El préstamo fue liquidado en su totalidad.
- Issued: El préstamo ha sido aprobado y emitido recientemente.
- Current: El préstamo se encuentra activo y los pagos se realizan conforme a lo establecido.
- In Grace Period: El préstamo se encuentra dentro del período de gracia permitido para regularizar pagos atrasados.
- Default: El préstamo ha entrado en incumplimiento de pago.
- Late (16-30 days): El préstamo presenta un retraso en los pagos de entre 16 y 30 días.
- Late (31-120 days): El préstamo presenta un retraso en los pagos de entre 31 y 120 días.
- Charged Off: El préstamo ha sido clasificado como incobrable por la entidad financiera.

Del análisis de los datos contenidos en la categoría 'loan_status' se observa lo siguiente:

- Los registros proporcionados describen el estado asociado a cada préstamo contenido en la base de datos.
- Los valores almacenados corresponden a categorías cualitativas y, por tanto, pueden ser representados mediante cadenas de texto.
- Se detectan 500 registros en los que el estado del préstamo no fue registrado adecuadamente, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.
- Considerar la creación de una categoría adicional que agrupe los estados asociados a incumplimiento o mora (por ejemplo, Default, Charged Off y categorías Late), en caso de que la variable vaya a emplearse posteriormente en tareas de análisis de riesgo o modelado predictivo.

### Categoría 'funded_amnt'

Esta categoría representa el monto total comprometido para el préstamo en el momento de su aprobación. El valor registrado puede diferir del monto inicialmente solicitado por el prestatario, ya que la entidad financiera puede realizar ajustes durante el proceso de evaluación crediticia. En consecuencia, esta variable refleja el importe efectivamente aprobado para el préstamo en el instante de su emisión.

In [ ]:
#caracteristicas_data[label_data[4]]
list(caracteristicas_data[label_data[4]].items())[:10]

{np.str_(''): np.int64(500),
 np.str_('1000.0'): np.int64(28),
 np.str_('10000.0'): np.int64(628),
 np.str_('10025.0'): np.int64(1),
 np.str_('10050.0'): np.int64(3),
 np.str_('10075.0'): np.int64(1),
 np.str_('10100.0'): np.int64(4),
 np.str_('10150.0'): np.int64(3),
 np.str_('10175.0'): np.int64(1),
 np.str_('10200.0'): np.int64(13),
 np.str_('10225.0'): np.int64(1),
 np.str_('10250.0'): np.int64(3),
 np.str_('10300.0'): np.int64(2),
 np.str_('10325.0'): np.int64(3),
 np.str_('10350.0'): np.int64(1),
 np.str_('10375.0'): np.int64(3),
 np.str_('10400.0'): np.int64(16),
 np.str_('10425.0'): np.int64(4),
 np.str_('10475.0'): np.int64(1),
 np.str_('10500.0'): np.int64(31),
 np.str_('10525.0'): np.int64(5),
 np.str_('10550.0'): np.int64(6),
 np.str_('10575.0'): np.int64(2),
 np.str_('10600.0'): np.int64(3),
 np.str_('10625.0'): np.int64(1),
 np.str_('10650.0'): np.int64(2),
 np.str_('10700.0'): np.int64(6),
 np.str_('10725.0'): np.int64(3),
 np.str_('10750.0'): np.int64(2),
 np.str_('1077

Del análisis de los datos contenidos en la categoría 'loan_amnt' se observa lo siguiente:

- Los registros proporcionados corresponden al monto de los préstamos aceptados que se encuentran asociados a cada solicitud contenida en la base de datos.
- Los valores almacenados representan cantidades numéricas y, por tanto, pueden ser tratados como datos de tipo flotante para facilitar su análisis y procesamiento.
- Se detectan 500 registros en los que el monto del préstamo no fue almacenado correctamente, encontrándose representado mediante una cadena vacía ('').
- No se identificaron valores con formatos atípicos o inconsistentes en comparación con el resto de los registros de esta categoría, lo que indica una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.
- Convertir los valores válidos de la categoría desde cadenas de texto a un tipo de dato numérico (por ejemplo, float), permitiendo la aplicación de técnicas estadísticas y de análisis cuantitativo sobre esta variable.

### Categoría 'term'

Esta categoría indica el plazo de amortización del préstamo, expresado en meses. 

Los valores posibles son 36 y 60 meses, correspondientes a los dos esquemas de financiamiento ofrecidos para los préstamos registrados en la base de datos. 

Esta variable permite identificar el horizonte temporal establecido para la liquidación de cada préstamo.

In [ ]:
#caracteristicas_data[label_data[5]]
list(caracteristicas_data[label_data[5]].items())[:10]

{np.str_(''): np.int64(500),
 np.str_(' 36 months'): np.int64(6281),
 np.str_(' 60 months'): np.int64(3219)}

Del análisis de los datos contenidos en la categoría 'term' se observa lo siguiente:

- Los registros proporcionados corresponden al plazo de amortización asociado a cada préstamo, expresado en meses.
- Los valores almacenados representan cantidades numéricas discretas y, por tanto, pueden ser tratados como datos de tipo entero. No obstante, es necesario eliminar la unidad de medida ("months") para facilitar su procesamiento y análisis.
- Se detectan 500 registros cuyos plazos no fueron almacenados correctamente, encontrándose representados mediante una cadena vacía ('').
- No se identificaron valores con formatos atípicos o inconsistentes en comparación con el resto de los registros de esta categoría, lo que evidencia una adecuada uniformidad en la representación de los datos.
- Los únicos plazos observados en los registros válidos corresponden a 36 y 60 meses.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.
- Eliminar la unidad de medida ("months") y convertir los valores válidos de la categoría desde cadenas de texto a un tipo de dato numérico entero (int).
- En caso de requerirse una representación más descriptiva de la información, conservar la unidad de medida mediante la documentación de la variable o la generación de una columna derivada para fines de presentación.

### Categoría 'int_rate'

Datos que se corresponden con la tasa de interés asociada al préstamo concedido al solicitante, expresada como un porcentaje sobre el monto financiado.

In [ ]:
#caracteristicas_data[label_data[6]]
list(caracteristicas_data[label_data[6]].items())[:10]

{np.str_(''): np.int64(500),
 np.str_('12.ìàé'): np.int64(48),
 np.str_('13.18'): np.int64(100),
 np.str_('13.33'): np.int64(418),
 np.str_('13.44'): np.int64(10),
 np.str_('13.66'): np.int64(56),
 np.str_('13.67'): np.int64(111),
 np.str_('13.99'): np.int64(402),
 np.str_('14.31'): np.int64(50),
 np.str_('14.33'): np.int64(33),
 np.str_('14.48'): np.int64(37),
 np.str_('14.65'): np.int64(306),
 np.str_('14.85'): np.int64(6),
 np.str_('14.99'): np.int64(39),
 np.str_('15.41'): np.int64(56),
 np.str_('15.59'): np.int64(50),
 np.str_('15.61'): np.int64(281),
 np.str_('15.77'): np.int64(6),
 np.str_('15.99'): np.int64(36),
 np.str_('16.49'): np.int64(26),
 np.str_('16.55'): np.int64(228),
 np.str_('16.59'): np.int64(5),
 np.str_('16.99'): np.int64(211),
 np.str_('17.14'): np.int64(25),
 np.str_('17.27'): np.int64(4),
 np.str_('17.57'): np.int64(222),
 np.str_('17.86'): np.int64(222),
 np.str_('17.97'): np.int64(5),
 np.str_('18.25'): np.int64(159),
 np.str_('18.49'): np.int64(31),
 np.str

In [107]:
valores_invalidos_rate = []
for key, _ in caracteristicas_data[label_data[6]].items():
    try:
        x = float(key)
    except:
        valores_invalidos_rate.append(key)
valores_invalidos_rate

[np.str_(''),
 np.str_('12.ìàé'),
 np.str_('18.ôåâ'),
 np.str_('24.ìàé'),
 np.str_('25.àâã'),
 np.str_('26.þíè'),
 np.str_('6.ìàð'),
 np.str_('9.àâã'),
 np.str_('àâã.18'),
 np.str_('àâã.19'),
 np.str_('àâã.38'),
 np.str_('àâã.49'),
 np.str_('àâã.67'),
 np.str_('äåê.29'),
 np.str_('äåê.39'),
 np.str_('äåê.59'),
 np.str_('äåê.69'),
 np.str_('äåê.88'),
 np.str_('äåê.99'),
 np.str_('ìàé.32'),
 np.str_('ìàé.93'),
 np.str_('íîå.22'),
 np.str_('íîå.44'),
 np.str_('íîå.48'),
 np.str_('íîå.49'),
 np.str_('íîå.53'),
 np.str_('íîå.99'),
 np.str_('îêò.49'),
 np.str_('îêò.64'),
 np.str_('îêò.78'),
 np.str_('îêò.99'),
 np.str_('ñåï.17'),
 np.str_('ñåï.49'),
 np.str_('ñåï.76'),
 np.str_('ñåï.99'),
 np.str_('þëè.26'),
 np.str_('þëè.49'),
 np.str_('þëè.89'),
 np.str_('þëè.91'),
 np.str_('þíè.24'),
 np.str_('þíè.39'),
 np.str_('þíè.49'),
 np.str_('þíè.68'),
 np.str_('þíè.89'),
 np.str_('þíè.92'),
 np.str_('þíè.99')]

Del análisis de los datos contenidos en la categoría 'int_rate' se observa lo siguiente:

- Los registros proporcionados corresponden a las tasas de interés establecidas a cada préstamo, expresada en porcentaje.
- Los valores almacenados representan cantidades numéricas y, por tanto, pueden ser tratados como datos de tipo flotante.
- Se detectan 500 registros cuyas tasas de interés no fueron almacenados correctamente, encontrándose representados mediante una cadena vacía ('').
- Se identificaron valores con formatos atípicos o inconsistentes en comparación con el resto de los registros de esta categoría, los cuales fueron agrupados dentro del arreglo valores_invalidos_rate.

Recomendaciones:

- Reemplazar las cadenas vacías ('') y los valores atípicos identificados por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

### Categoría 'installment'

En esta categoría se agrega el pago mensual que deberá realizar el prestatario en caso de que el préstamo sea otorgado.

In [ ]:
#caracteristicas_data[label_data[7]]
list(caracteristicas_data[label_data[7]].items())[:10]

{np.str_(''): np.int64(500),
 np.str_('100.04'): np.int64(1),
 np.str_('100.06'): np.int64(6),
 np.str_('100.21'): np.int64(1),
 np.str_('100.5'): np.int64(1),
 np.str_('100.64'): np.int64(11),
 np.str_('100.68'): np.int64(1),
 np.str_('1000.37'): np.int64(1),
 np.str_('1000.59'): np.int64(4),
 np.str_('1001.15'): np.int64(3),
 np.str_('1002.03'): np.int64(1),
 np.str_('1002.9'): np.int64(1),
 np.str_('1004.91'): np.int64(3),
 np.str_('1005.43'): np.int64(1),
 np.str_('1006.35'): np.int64(4),
 np.str_('1006.68'): np.int64(3),
 np.str_('1007.52'): np.int64(1),
 np.str_('1009.09'): np.int64(1),
 np.str_('101.35'): np.int64(1),
 np.str_('101.56'): np.int64(3),
 np.str_('101.73'): np.int64(1),
 np.str_('1010.68'): np.int64(1),
 np.str_('1013.11'): np.int64(1),
 np.str_('1015.6'): np.int64(3),
 np.str_('1015.79'): np.int64(1),
 np.str_('102.02'): np.int64(1),
 np.str_('102.06'): np.int64(1),
 np.str_('102.52'): np.int64(2),
 np.str_('102.57'): np.int64(1),
 np.str_('102.91'): np.int64(1),
 

Del análisis de los datos contenidos en la categoría 'installment' se observa lo siguiente:

- Los registros proporcionados corresponden al pago mensual que debe efectuar cada cliente.
- Los valores almacenados representan cantidades numéricas y, por tanto, pueden ser tratados como datos de tipo flotante para facilitar su análisis y procesamiento.
- Se detectan 500 registros en los que el monto del préstamo no fue almacenado correctamente, encontrándose representado mediante una cadena vacía ('').
- No se identificaron valores con formatos atípicos o inconsistentes en comparación con el resto de los registros de esta categoría, lo que indica una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.
- Convertir los valores válidos de la categoría desde cadenas de texto a un tipo de dato numérico (por ejemplo, float), permitiendo la aplicación de técnicas estadísticas y de análisis cuantitativo sobre esta variable.

### Categoría 'grade'

Calificación del préstamo asignada por Lending Club (LC).

In [ ]:
#caracteristicas_data[label_data[8]]
list(caracteristicas_data[label_data[8]].items())[:10]

{np.str_(''): np.int64(515),
 np.str_('A'): np.int64(1632),
 np.str_('B'): np.int64(2606),
 np.str_('C'): np.int64(2766),
 np.str_('D'): np.int64(1389),
 np.str_('E'): np.int64(816),
 np.str_('F'): np.int64(236),
 np.str_('G'): np.int64(40)}

Los registros contenidos en esta categoría corresponden a la calificación crediticia asignada al préstamo por la entidad financiera. Dicha calificación se expresa mediante categorías ordenadas de la letra A a la G, donde las categorías más cercanas a A representan un menor nivel de riesgo crediticio, mientras que las categorías próximas a G indican un mayor riesgo asociado al préstamo.

Del análisis de los datos contenidos en la categoría 'grade' se observa lo siguiente:

- Los registros proporcionados describen la clasificación de cada uno de los préstamos efectuados por la entidad.
- Los valores almacenados corresponden a categorías cualitativas y, por tanto, pueden ser representados mediante cadenas de texto.
- Se detectan 515 registros en los que la categoría no está definida de manera adecuada, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

### Categoría 'sub_grade'

Calificación del préstamo asignada por Lending Club (LC). (clasificación basada en sub-grados)

In [173]:
#caracteristicas_data[label_data[9]]
list(caracteristicas_data[label_data[9]].items())[:10]

[(np.str_(''), np.int64(514)),
 (np.str_('A1'), np.int64(285)),
 (np.str_('A2'), np.int64(278)),
 (np.str_('A3'), np.int64(239)),
 (np.str_('A4'), np.int64(323)),
 (np.str_('A5'), np.int64(502)),
 (np.str_('B1'), np.int64(509)),
 (np.str_('B2'), np.int64(517)),
 (np.str_('B3'), np.int64(530)),
 (np.str_('B4'), np.int64(553))]

Esta variable representa la subcalificación crediticia asignada al préstamo por la entidad financiera. La clasificación se expresa mediante una combinación de una letra y un número, donde la letra identifica la categoría general de riesgo y el número establece una graduación más específica dentro de dicha categoría. Esta variable permite una caracterización más detallada del perfil de riesgo asociado al préstamo.

Del análisis de los datos contenidos en la categoría 'sub_grade' se observa lo siguiente:

- Los registros proporcionados describen la clasificación de cada uno de los préstamos efectuados por la entidad por medio de sub-grados identificados por la asignación numérica de la clasificación.
- Los valores almacenados corresponden a categorías cualitativas y, por tanto, pueden ser representados mediante cadenas de texto.
- Se detectan 514 registros en los que la categoría no está definida de manera adecuada, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

### Categoría 'verification_status'

Indica el estado de verificación de los ingresos declarados por el solicitante, especificando si dichos ingresos fueron verificados por la entidad financiera, no fueron verificados o si únicamente se verificó la fuente de ingresos.

In [174]:
#caracteristicas_data[label_data[10]]
list(caracteristicas_data[label_data[10]].items())[:10]

[(np.str_(''), np.int64(500)),
 (np.str_('Not Verified'), np.int64(2673)),
 (np.str_('Source Verified'), np.int64(4116)),
 (np.str_('Verified'), np.int64(2711))]

Los registros contenidos en esta categoría corresponden al estado de verificación de los ingresos declarados por el prestatario durante el proceso de solicitud del préstamo. En particular, la variable distingue entre ingresos no verificados (Not Verified), ingresos cuya fuente fue verificada (Source Verified) e ingresos completamente verificados por la entidad financiera (Verified).

Del análisis de los datos contenidos en la categoría 'verification_status' se observa lo siguiente:

- Los registros proporcionados describen el estado de verificación de los ingresos del cliente que solicitó el préstamo a la entidad.
- Los valores almacenados corresponden a categorías cualitativas y, por tanto, pueden ser representados mediante cadenas de texto.
- Se detectan 500 registros en los que la categoría no está definida de manera adecuada, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

### Categoría 'url'

URL de la página de Lending Club que contiene la información asociada a la solicitud o publicación del préstamo.

In [175]:
#caracteristicas_data[label_data[11]]
list(caracteristicas_data[label_data[11]].items())[:10]

[(np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=12606806'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=13026045'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=1312426'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=14187903'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=14428496'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=14470171'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=14550531'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=14669896'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=15389420'),
  np.int64(1)),
 (np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=16130931'),
  np.int64(1))]

In [141]:
arreglo_count = [item for _, item in caracteristicas_data[label_data[11]].items()]
max(arreglo_count)

np.int64(1)

El análisis de las direcciones contenidas en la categoría 'url' evidencia que:

- Mantienen una correspondencia única con cada uno de los registros extraídos, por lo que pueden considerarse direcciones electrónicas únicas dentro del conjunto de datos, mediante las cuales se extrajeron los datos que fueron registrados.
- Se descarta la presencia de registros duplicados, ya que cada dirección posee una única ocurrencia en la base de datos analizada.

### Categoría 'addr_state'

El estado (entidad federativa) reportado por el prestatario en la solicitud del préstamo.

In [176]:
#caracteristicas_data[label_data[12]]
list(caracteristicas_data[label_data[12]].items())[:10]

[(np.str_(''), np.int64(500)),
 (np.str_('AK'), np.int64(26)),
 (np.str_('AL'), np.int64(119)),
 (np.str_('AR'), np.int64(74)),
 (np.str_('AZ'), np.int64(220)),
 (np.str_('CA'), np.int64(1336)),
 (np.str_('CO'), np.int64(201)),
 (np.str_('CT'), np.int64(143)),
 (np.str_('DC'), np.int64(27)),
 (np.str_('DE'), np.int64(27))]

Los registros contenidos en esta categoría corresponden al estado de residencia declarado por el prestatario durante el proceso de solicitud del préstamo. La información se encuentra representada mediante abreviaturas oficiales de los estados y territorios de los Estados Unidos, por lo que la variable proporciona una referencia geográfica asociada a cada solicitante.

Del análisis de los datos contenidos en la categoría 'addr_state' se observa lo siguiente:

- Los registros proporcionados describen el estado de la persona que solicitó el préstamo.
- Los valores almacenados corresponden a categorías cualitativas y, por tanto, pueden ser representados mediante cadenas de texto.
- Se detectan 500 registros en los que la categoría no está definida de manera adecuada, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

### Categoría 'total_pymnt'

Representa el monto acumulado que el prestatario ha reembolsado hasta la fecha respecto al capital financiado, incluyendo los pagos efectuados durante la vigencia del préstamo.

In [177]:
#caracteristicas_data[label_data[13]]
list(caracteristicas_data[label_data[13]].items())[:10]

[(np.str_(''), np.int64(500)),
 (np.str_('0.0'), np.int64(384)),
 (np.str_('0.74'), np.int64(1)),
 (np.str_('100.0'), np.int64(1)),
 (np.str_('1000.62'), np.int64(1)),
 (np.str_('1000.81'), np.int64(2)),
 (np.str_('10002.6'), np.int64(1)),
 (np.str_('10006.27'), np.int64(1)),
 (np.str_('1001.69'), np.int64(1)),
 (np.str_('1001.71'), np.int64(1))]

Los registros contenidos en esta categoría corresponden al pago efectuado por los clientes en el momento en el cual se realizó la recopilación de los datos.

Del análisis de los datos contenidos en la categoría 'total_pymnt' se observa lo siguiente:

- Los registros proporcionados describen el pago efectuado por los clientes.
- Los valores almacenados corresponden a categorías cuantitativas y, por tanto, pueden ser representados mediante valores de punto flotante.
- Se detectan 500 registros en los que la categoría no está definida de manera adecuada, encontrándose representado mediante una cadena vacía ('').
- No se identificaron categorías redundantes, formatos inconsistentes ni etiquetas atípicas en comparación con el resto de los registros, lo que evidencia una adecuada uniformidad en la representación de los datos.

Recomendaciones:

- Reemplazar las cadenas vacías ('') por el valor np.nan, con el fin de normalizar la representación de datos faltantes y facilitar su identificación, tratamiento y análisis posterior.

## Casteo de datos

In [115]:
raw_data_copy = raw_data_np.copy()
casted_data_np = np.empty_like(raw_data_copy[:,1:], dtype=object)

for i in range(raw_data_np.shape[0]):
    for j in range(1, raw_data_np.shape[1]):
        try:
            if raw_data_copy[i, 0] != 'id':
                casted_data_np[i, j-1] = float(raw_data_copy[i, j])
            else:
                casted_data_np[i, j-1] = raw_data_copy[i, j]
        except ValueError:
            casted_data_np[i, j-1] = raw_data_copy[i, j]
casted_data_np

array([[np.str_('48010226'), np.str_('57693261'), np.str_('59432726'), ..., np.str_('50415990'),
        np.str_('46154151'), np.str_('66055249')],
       [np.str_('May-15'), np.str_(''), np.str_('Sep-15'), ..., np.str_('Jun-15'),
        np.str_('Apr-15'), np.str_('Dec-15')],
       [35000.0, 30000.0, 15000.0, ..., 10000.0, np.str_(''), 10000.0],
       ...,
       [np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226'),
        np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261'),
        np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726'), ...,
        np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990'),
        np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151'),
        np.str_('https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249')],
       [np.str_('CA'), np.str_('NY'), np.str_('PA'), ..., np.str_('CA'), np.str_('OH'),
    

## Definición y Corrección de la Estructura de Datos

### Estructuración del diccionario de categorías

In [117]:
data_dict = {}
for i in range(len(raw_data_np[:,0])):
    data_dict[raw_data_np[i,0]] = casted_data_np[i]
data_dict

{np.str_('id'): array([np.str_('48010226'), np.str_('57693261'), np.str_('59432726'), ..., np.str_('50415990'),
        np.str_('46154151'), np.str_('66055249')], shape=(10000,), dtype=object),
 np.str_('issue_d'): array([np.str_('May-15'), np.str_(''), np.str_('Sep-15'), ..., np.str_('Jun-15'),
        np.str_('Apr-15'), np.str_('Dec-15')], shape=(10000,), dtype=object),
 np.str_('loan_amnt'): array([35000.0, 30000.0, 15000.0, ..., 10000.0, np.str_(''), 10000.0], shape=(10000,), dtype=object),
 np.str_('loan_status'): array([np.str_('Current'), np.str_('Current'), np.str_('Current'), ..., np.str_('Current'),
        np.str_('Current'), np.str_('Current')], shape=(10000,), dtype=object),
 np.str_('funded_amnt'): array([35000.0, 30000.0, 15000.0, ..., 10000.0, 10000.0, 10000.0], shape=(10000,), dtype=object),
 np.str_('term'): array([np.str_(' 36 months'), np.str_(' 36 months'), np.str_(' 36 months'), ...,
        np.str_(' 36 months'), np.str_(' 36 months'), np.str_(' 36 months')],
   

### Corrección de los elementos integrados a la categoría 'issue_d'

In [118]:
data_dict['issue_d'] = np.where(
    data_dict['issue_d'] == '',
    np.nan,
    data_dict['issue_d']
)
registros_faltantes = np.sum(pd.isna(data_dict['issue_d']))
print(registros_faltantes)

500


In [119]:
total_registros = np.size(data_dict['issue_d'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'issue_d', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'loan_amnt'

In [120]:
data_dict['loan_amnt'] = np.where(
    data_dict['loan_amnt'] == '',
    np.nan,
    data_dict['loan_amnt']
)
registros_faltantes = np.sum(pd.isna(data_dict['loan_amnt']))
print(registros_faltantes)

500


In [121]:
total_registros = np.size(data_dict['loan_amnt'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'loan_amnt', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'loan_status'

In [122]:
data_dict['loan_status'] = np.where(
    data_dict['loan_status'] == '',
    np.nan,
    data_dict['loan_status']
)
registros_faltantes = np.sum(pd.isna(data_dict['loan_status']))
print(registros_faltantes)

500


In [123]:
total_registros = np.size(data_dict['loan_status'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'loan_status', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'funded_amnt'

In [124]:
data_dict['funded_amnt'] = np.where(
    data_dict['funded_amnt'] == '',
    np.nan,
    data_dict['funded_amnt']
)
registros_faltantes = np.sum(pd.isna(data_dict['funded_amnt']))
print(registros_faltantes)

500


In [125]:
total_registros = np.size(data_dict['funded_amnt'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'funded_amnt', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'term'

In [126]:
for i in range(len(data_dict['term'])):
    if data_dict['term'][i] != "":
        data_dict['term'][i] = int((data_dict['term'][i].split()[0]))
    else:
        data_dict['term'][i] = np.nan
type(data_dict['term'])

numpy.ndarray

In [127]:
registros_faltantes = np.sum(pd.isna(data_dict['term']))
print(registros_faltantes)

500


In [128]:
total_registros = np.size(data_dict['term'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevaron a cabo dos procesos de transformación sobre los datos contenidos en esta categoría:

- Los registros válidos fueron convertidos a valores enteros, ya que representan la cantidad de meses correspondiente al plazo del préstamo otorgado.
- Los registros vacíos fueron reemplazados por valores np.nan con el fin de identificar explícitamente los datos faltantes.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado evidencia un alto nivel de completitud en la información asociada a esta variable. No obstante, la presencia de datos faltantes deberá ser considerada en las etapas posteriores de preprocesamiento y análisis para garantizar la validez de los resultados obtenidos.

### Corrección de los elementos integrados a la categoría 'int_rate'

In [129]:
data_dict['int_rate'] = np.where(
    np.isin(data_dict['int_rate'], valores_invalidos_rate),
    np.nan,
    data_dict['int_rate']
)
registros_faltantes = np.sum(pd.isna(data_dict['int_rate']))
print(registros_faltantes)

6004


In [130]:
total_registros = np.size(data_dict['int_rate'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

60.040000000000006


Se llevó a cabo un proceso de depuración de los datos registrados en esta categoría, encontrándose que el 60.04 % de los registros presentaban un formato inadecuado. Este resultado evidencia una baja calidad de los datos asociados a la variable, lo que dificulta la aplicación de procedimientos de corrección sin introducir un sesgo significativo en los resultados o comprometer la representatividad de la información original.

En consecuencia, se determinó que los datos contenidos en esta columna no poseen la confiabilidad necesaria para ser considerados en etapas posteriores del análisis. Por ello, la variable fue descartada del estudio, imposibilitando la evaluación de aspectos como la influencia de la tasa de interés sobre la capacidad o voluntad de pago de los clientes de la institución financiera.

### Corrección de los elementos integrados a la categoría 'installment'

In [131]:
data_dict['installment'] = np.where(
    data_dict['installment'] == '',
    np.nan,
    data_dict['installment']
)
registros_faltantes = np.sum(pd.isna(data_dict['installment']))
print(registros_faltantes)

500


In [132]:
total_registros = np.size(data_dict['int_rate'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'installment', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'grade'

In [133]:
data_dict['grade'] = np.where(
    data_dict['grade'] == '',
    np.nan,
    data_dict['grade']
)
registros_faltantes = np.sum(pd.isna(data_dict['grade']))
print(registros_faltantes)

515


In [134]:
total_registros = np.size(data_dict['grade'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.1499999999999995


Se llevó a cabo la corrección de los registros contenidos en la categoría 'installment', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5.15 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'sub_grade'

In [139]:
data_dict['sub_grade'] = np.where(
    data_dict['sub_grade'] == '',
    np.nan,
    data_dict['sub_grade']
)
registros_faltantes = np.sum(pd.isna(data_dict['sub_grade']))
print(registros_faltantes)

514


In [140]:
total_registros = np.size(data_dict['sub_grade'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.140000000000001


Se llevó a cabo la corrección de los registros contenidos en la categoría 'installment', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5.14 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'verification_status'

In [142]:
data_dict['verification_status'] = np.where(
    data_dict['verification_status'] == '',
    np.nan,
    data_dict['verification_status']
)
registros_faltantes = np.sum(pd.isna(data_dict['verification_status']))
print(registros_faltantes)

500


In [143]:
total_registros = np.size(data_dict['verification_status'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'verification_status', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5.00 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'addr_state'

In [144]:
data_dict['addr_state'] = np.where(
    data_dict['addr_state'] == '',
    np.nan,
    data_dict['addr_state']
)
registros_faltantes = np.sum(pd.isna(data_dict['addr_state']))
print(registros_faltantes)

500


In [145]:
total_registros = np.size(data_dict['addr_state'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'verification_status', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5.00 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

### Corrección de los elementos integrados a la categoría 'total_pymnt'

In [146]:
data_dict['total_pymnt'] = np.where(
    data_dict['total_pymnt'] == '',
    np.nan,
    data_dict['total_pymnt']
)
registros_faltantes = np.sum(pd.isna(data_dict['total_pymnt']))
print(registros_faltantes)

500


In [149]:
total_registros = np.size(data_dict['total_pymnt'])
pct_b = registros_faltantes/total_registros*100
print(pct_b)

5.0


Se llevó a cabo la corrección de los registros contenidos en la categoría 'verification_status', reemplazando las etiquetas vacías ('') por el valor np.nan, con el propósito de estandarizar la representación de datos faltantes dentro de la estructura de datos.

Posteriormente, se calculó la proporción de valores faltantes respecto al total de registros disponibles, obteniéndose un porcentaje del 5.00 %. Este resultado indica que la completitud de la información asociada a esta variable es elevada, aunque la presencia de datos faltantes deberá considerarse durante las etapas posteriores de análisis y preprocesamiento.

## Splitting the Dataset

### Splitting the Columns

### Re-importing the Dataset

### The Names of the Columns

## Creating Checkpoints:

In [ ]:
def checkpoint(file_name, checkpoint_header, checkpoint_data):
    np.savez(file_name, header = checkpoint_header, data = checkpoint_data)
    checkpoint_variable = np.load(file_name + ".npz")
    return(checkpoint_variable)

## Manipulating String Columns

### Issue Date

In [ ]:
months = np.array(['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])

In [ ]:
for i in range(13):
        loan_data_strings[:,0] = np.where(loan_data_strings[:,0] == months[i],
                                          i,
                                          loan_data_strings[:,0])

### Loan Status

### Term

### Grade and Subgrade

#### Filling Sub Grade

In [ ]:
for i in np.unique(loan_data_strings[:,3])[1:]:
    loan_data_strings[:,4] = np.where((loan_data_strings[:,4] == '') & (loan_data_strings[:,3] == i),
                                      i + '5',
                                      loan_data_strings[:,4])

#### Removing Grade

#### Converting Sub Grade

In [ ]:
keys = list(np.unique(loan_data_strings[:,3]))                         
values = list(range(1, np.unique(loan_data_strings[:,3]).shape[0] + 1)) 
dict_sub_grade = dict(zip(keys, values))

In [ ]:
for i in np.unique(loan_data_strings[:,3]):
        loan_data_strings[:,3] = np.where(loan_data_strings[:,3] == i, 
                                          dict_sub_grade[i],
                                          loan_data_strings[:,3])

### Verification Status

### URL

### State Address

In [ ]:
states_west = np.array(['WA', 'OR','CA','NV','ID','MT', 'WY','UT','CO', 'AZ','NM','HI','AK'])
states_south = np.array(['TX','OK','AR','LA','MS','AL','TN','KY','FL','GA','SC','NC','VA','WV','MD','DE','DC'])
states_midwest = np.array(['ND','SD','NE','KS','MN','IA','MO','WI','IL','IN','MI','OH'])
states_east = np.array(['PA','NY','NJ','CT','MA','VT','NH','ME','RI'])

https://www2.census.gov/geo/pdfs/maps-data/maps/reference/us_regdiv.pdf

## Converting to Numbers

### Checkpoint 1: Strings

## Manipulating Numeric Columns

### Substitute "Filler" Values

#### ID

#### Temporary Stats

#### Funded Amount

In [ ]:
loan_data_numeric[:,2] = np.where(loan_data_numeric[:,2] == temporary_fill, 
                                  temporary_stats[0, columns_numeric[2]],
                                  loan_data_numeric[:,2])
loan_data_numeric[:,2]

#### Loaned Amount, Interest Rate, Total Payment, Installment

### Currency Change

#### The Exchange Rate

In [ ]:

for i in range(1,13):
    exchange_rate = np.where(exchange_rate == i,
                             EUR_USD[i-1],
                             exchange_rate)    


#### From USD to EUR

In [ ]:
for i in columns_dollar:
    loan_data_numeric = np.hstack((loan_data_numeric, np.reshape(loan_data_numeric[:,i] / loan_data_numeric[:,6], (10000,1))))

#### Expanding the header

In [ ]:
header_additional = np.array([column_name + '_EUR' for column_name in header_numeric[columns_dollar]])

In [ ]:
columns_index_order = [0,1,7,2,8,3,4,9,5,10,6]

### Interest Rate

### Checkpoint 2: Numeric

In [ ]:
checkpoint_numeric['header'], checkpoint_numeric['data']

## Creating the "Complete" Dataset

## Sorting the New Dataset

## Storing the New Dataset